plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/organize/imports/ImportsComputer.xtend

```diff
@@ -145,271 +148,216 @@ public class ImportsComputer {
-			// The first Identifiable in there is bound to the first thing the scoping encountered. For the time being, lets take any:
-			result += multimap.get(name).get(0)
+		// potential match via namespace
+		if (usedName.contains(".")) {
+			val segments = usedName.split("\\.")
+			// 2 segments, potential namespace access
+			if (segments.size == 2 && (segments.last == qName.lastSegment))
+				return true
 		}
 
-		return result;
+		return qName.lastSegment == EXPORT_DEFAULT_NAME && qName.segmentCount > 1 &&
+			qName.getSegment(qName.segmentCount - 2) == usedName
 	}
 
 	/**
-	 * Depending on interaction-mode:
-	 * Present a user dialog and let the user choose a distinct import for each unresolved problem.
-	 *
-	 * For each key in {@code multiMap} exactly one solution is returned.
-	 *
+	 * In most cases we cannot compute imports from proxy objects, as essential information like declared name, exported name etc.
+	 * are not available. If the provided object is a proxy, tries to resolve it. If object is still a proxy logs a warning. In either case returns result of resolving a proxy
+	 * @returns resolved EObject or proxy
 	 */
-	private def <T> List<T> disambiguate(Multimap<String, T> multiMapName2Candidates,
-		Interaction interaction) throws BreakException {
-		// for each name exactly one solution must be picked:
-		val result = <T>newArrayList();
-		if (multiMapName2Candidates.empty) return result;
-
-		switch (interaction) {
-			case breakBuild: {
-				throw new BreakException("Cannot automatically disambiguate the imports of " +
-					multiMapName2Candidates.keySet.toList)
-			}
-			case takeFirst: {
-				return takefirst(multiMapName2Candidates);
+	private def EObject tryGetEObjectOrProxy(EObject eobject, Resource contextResource) {
+		var eo = eobject
+		if (eo.eIsProxy) {
+			eo = EcoreUtil.resolve(eo, contextResource)
+			var data = "";
+			if (eo.eIsProxy) {
+				if (eo instanceof TClass) {
+					data = eo.name + " "
+				}
+				logger.warn("Cannot resolve proxy " + data + eo)
 			}
-			case queryUser: {
-			} // follows
 		}
+		return eo;
+	}
 
-		val ILabelProvider labelProvider = importProvidedElementLabelprovider;
+	/**
+	 * Checks if the provided uri does not point to a module. For {@link IEObjectDescription} instances
+	 * the URI for modules (files) will be in form {@code some/path/1} while for the AST elements inside the module
+	 * it will be {@code some/path/1/@topLevelTypes.4} or {@code some/path/1/@variables.7}
+	 */
+	private def boolean isNotModule(URI uri) {
+		val res = uri?.fragment.endsWith("/1")
+		return !res
+	}
 
-		val Object[][] openChoices = Multimaps3.createOptions(multiMapName2Candidates);
+	private def boolean isVisible(IEObjectDescription ieod, Resource contextResource) {
+		val eo = tryGetEObjectOrProxy(ieod.EObjectOrProxy, contextResource)
+		if (eo instanceof Type && isSubtypeOfType(eo.eClass))
```

Comment:
Are these `isSubtypeOfType` ... necessary??

Reply:
good question. Actually it is copy from scopes. I wanted to preserve the logic.

Reply:
I don't see any need really. The `instanceof`clause already checks that...